Formula/ecflow-ui.rb

```diff
@@ -0,0 +1,39 @@
+class EcflowUi < Formula
+  desc "User interface for client/server workflow package"
+  homepage "https://confluence.ecmwf.int/display/ECFLOW"
+  url "https://confluence.ecmwf.int/download/attachments/8650755/ecFlow-5.8.1-Source.tar.gz"
+  sha256 "be2ff0b76da2917a00ba13cd02b2992616e374b102ea92dff63a022fbdaa1884"
+  license "Apache-2.0"
+
+  depends_on "boost" => :build
+  depends_on "cmake" => :build
+  depends_on "openssl@1.1"
+  depends_on "qt"
+
+  # requires C++17 compiler to build with Qt
+  fails_with gcc: "5"
+
+  def install
+    mkdir "build" do
+      system "cmake", "..",
+                      "-DENABLE_SSL=1",
+                      "-DENABLE_PYTHON=OFF",
+                      "-DECBUILD_LOG_LEVEL=DEBUG",
+                      "-DENABLE_SERVER=OFF",
+                      *std_cmake_args
+      system "make", "install"
+    end
+  end
+
+  # current tests assume the existence of ecflow_client, but we may not always supply
+  # this in future versions, but for now it's the best test we can do to make sure things
+  # are linked properly
+  test do
+    system "#{bin}/ecflow_client", "--help"
+    help_out = shell_output("#{bin}/ecflow_client --help")
+    assert_match "Ecflow version", help_out
+    assert_match "ecflow_client", help_out
+    assert_match "begin", help_out
+    assert_match "zombie_get", help_out
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
Hi @SMillerDev thanks for the feedback (all very useful!). I did read that advice when writing the formula, but ecflow-ui is a purely GUI application and has no meaningful command-line test. That said, I just spoke to the main developer and we agreed that we should be able to add something similar to what the advice suggests, i.e. have a mode where it starts without GUI, reads and prints config info. This would be included in the next release, but that is not expected for a few weeks at least. What do you think about releasing it already and I'll update the testing with the next release?

Reply:
People have lived without the formula so far, they can wait a couple more weeks.